### PR TITLE
Clean up constructors Arrow Core Datatypes.

### PR DESCRIPTION
**Pull request by nomisRev**
This PR deprecates up all duplicated constructors, and deprecates all `just` constructors.

This leaves behind only the actual constructors, and the extension constructors.

All constructors of Either after removal of deprecated methods:

```
val x: Either<String, Int> = Either.Right(1)
val y: Either<String, Int> = Either.Left("one")
val x2: Either<String, Int> = 1.right()
val y2: Either<String, Int> = "one".left()
```

### CODE REVIEW

**Review by franciscodr — APPROVED** [review]
Thanks @nomisRev!

File: arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 1]
```diff
@@ -431,6 +431,14 @@ sealed class Option<out A> : OptionOf<A> {
...
+      "empty is deprecated, and will be removed in 0.13.0. Please use None instead.",
+      ReplaceWith(
+        "None",
+        "arrow.core.None"
+      ),
+      DeprecationLevel.WARNING
+    )
     fun <A> empty(): Option<A> = None
 
     val unit: Option<Unit> = Some(Unit)
```
Comment (franciscodr) [inline arrow-core-data/src/main/kotlin/arrow/core/Option.kt:444]: I think we also discussed removing the `unit` constructor method, didn't we?

Comment (nomisRev) [reply arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: Yes, you're correct @franciscodr. I updated the PR. I checked against 0.11.0.3 and most of these were added after, so we could get away with removing them insteaad of deprecating 😄

Thanks for the reminder! 🕵️

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt [hunk 2]
```diff
@@ -183,7 +183,7 @@ sealed class Eval<out A> : EvalOf<A> {
     fun raise(t: Throwable): Eval<Nothing> =
       defer { throw t }
 
-    val unit: Eval<Unit> = Now(kotlin.Unit)
+    val Unit: Eval<Unit> = Now(kotlin.Unit)
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:186]: This was uppercased on `0.11.0`

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 3]
```diff
@@ -346,10 +356,19 @@ class NonEmptyList<out A>(
     fun <A> fromListUnsafe(l: List<A>): NonEmptyList<A> =
       NonEmptyList(l)
 
+    @Deprecated(
+      "just is deprecated, and will be removed in 0.13.0. Please use NonEmptyList.of instead.",
+      ReplaceWith(
+        "NonEmptyList.of(a)",
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:362]:
```suggestion
        "nonEmptyListOf(a)",
```

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt (hunk 2 above)
Comment (i-walker) [reply arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:186]: Do we want those to be upper-cased or lowercased in Arrow for other libs and API's?

**Review by i-walker — APPROVED** [review]
Thanks @nomisRev 🙌 , left a general question below on naming conventions.

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt (hunk 2 above)
Comment (nomisRev) [reply arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:186]: Looking at the other (smart) constructors in Arrow, I think we should deprecate all of these in favor of their right hand side implementations. 

```
val Unit: Eval<Unit> = Now(kotlin.Unit)
val True: Eval<Boolean> = Now(true)
val False: Eval<Boolean> = Now(false)
val Zero: Eval<Int> = Now(0)
val One: Eval<Int> = Now(1)
```

If all agree I can add this change in this PR as well.

Comment (i-walker) [reply arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:193]: I am sorry, I saw the reply after merging it. 
Let's move that to a different PR 😄

**Pull request merged**